octopusdeploy/schema_deployment_step.go

```diff
@@ -42,60 +44,41 @@ func expandDeploymentStep(flattenedStep map[string]interface{}) *deployments.Dep
-			step.Actions = append(step.Actions, action)
-		}
-	}
-
-	if v, ok := flattenedStep["apply_terraform_template_action"]; ok {
-		for _, tfAction := range v.([]interface{}) {
-			action := expandApplyTerraformTemplateAction(tfAction.(map[string]interface{}))
-			step.Actions = append(step.Actions, action)
-		}
-	}
+	var sort_order map[string]int = make(map[string]int)
 
-	if v, ok := flattenedStep["deploy_package_action"]; ok {
-		for _, tfAction := range v.([]interface{}) {
-			action := expandDeployPackageAction(tfAction.(map[string]interface{}))
-			step.Actions = append(step.Actions, action)
-		}
-	}
+	step_expansion := func(step_type_name string, step_type_action func(map[string]interface{}) *deployments.DeploymentAction) {
+		if v, ok := flattenedStep[step_type_name]; ok {
+			for _, tfAction := range v.([]interface{}) {
+				flattenedAction := tfAction.(map[string]interface{})
+				action := step_type_action(flattenedAction)
+				step.Actions = append(step.Actions, action)
 
-	if v, ok := flattenedStep["deploy_windows_service_action"]; ok {
-		for _, tfAction := range v.([]interface{}) {
-			action := expandDeployWindowsServiceAction(tfAction.(map[string]interface{}))
-			step.Actions = append(step.Actions, action)
-		}
-	}
-
-	if v, ok := flattenedStep["run_script_action"]; ok {
-		for _, tfAction := range v.([]interface{}) {
-			action := expandRunScriptAction(tfAction.(map[string]interface{}))
-			step.Actions = append(step.Actions, action)
+				// Pull out the sort_order if it exists. This is used to sort the actions later
+				if posn, ok := flattenedAction["sort_order"].(int); ok && posn >= 0 {
+					name := flattenedAction["name"].(string)
+					sort_order[name] = posn
+				}
+			}
 		}
 	}
 
-	if v, ok := flattenedStep["run_kubectl_script_action"]; ok {
-		for _, tfAction := range v.([]interface{}) {
-			action := expandRunKubectlScriptAction(tfAction.(map[string]interface{}))
-			step.Actions = append(step.Actions, action)
-		}
-	}
+	step_expansion("action", expandAction)
+	step_expansion("manual_intervention_action", expandManualInterventionAction)
+	step_expansion("apply_terraform_template_action", expandApplyTerraformTemplateAction)
+	step_expansion("deploy_package_action", expandDeployPackageAction)
+	step_expansion("deploy_windows_service_action", expandDeployWindowsServiceAction)
+	step_expansion("run_script_action", expandRunScriptAction)
+	step_expansion("run_kubectl_script_action", expandRunKubectlScriptAction)
+	step_expansion("deploy_kubernetes_secret_action", expandDeployKubernetesSecretAction)
 
-	if v, ok := flattenedStep["deploy_kubernetes_secret_action"]; ok {
-		for _, tfAction := range v.([]interface{}) {
-			action := expandDeployKubernetesSecretAction(tfAction.(map[string]interface{}))
-			step.Actions = append(step.Actions, action)
+	// Now that we have extracted all the steps off each of the properties into a single array, sort the array by the sort_order if provided
```

Comment:
so this does not seem to validate that the user hasn't made a mistake a typed the same order number twice?

Reply:
Yeah I had considered doing some validation around that. Rather than failing its easier to just log it with a warning\message. 
The current terraform provider library is actually really bad at allowing complex validation prior to apply. Im hoping the new one fixes some of these things up.

Reply:
Ill at least put a log around it. Its a shame there isnt simple linq-style syntax in go :)